src/breaker/SamplingBreaker.ts

```diff
@@ -0,0 +1,172 @@
+import { CircuitState } from '../CircuitBreakerPolicy';
+import { IBreaker } from './Breaker';
+
+interface IWindow {
+  startedAt: number;
+  failures: number;
+  successes: number;
+}
+
+export interface ISamplingBreakerOptions {
+  /**
+   * Percentage (from 0 to 1) of requests that need to fail before we'll
+   * open the circuit.
+   */
+  threshold: number;
+
+  /**
+   * Length of time over which to sample.
+   */
+  duration: number;
+
+  /**
+   * Minimum number of RPS needed to be able to (potentially) open the circuit.
+   * Useful to avoid unnecessarily tripping under low load.
+   */
+  minimumRps?: number;
+}
+
+/**
+ * SamplingBreaker breaks if more than `threshold` percentage of calls over the
+ * last `samplingDuration`, so long as there's at least `minimumRps` (to avoid
+ * closing unnecessarily under low RPS).
+ */
+export class SamplingBreaker implements IBreaker {
+  private readonly threshold: number;
+  private readonly minimumRpms: number;
+  private readonly duration: number;
+  private readonly windowSize: number;
+
+  private windows: IWindow[] = [];
+  private currentWindow = 0;
+  private currentFailures = 0;
+  private currentSuccesses = 0;
+
+  /**
+   * Gets a snapshot of internal breaker state. Used for tests.
+   */
+  public get state() {
```

Comment:
This is somewhat of a hot take, but an API shouldn't expose internal state for the sake of testing, this is an anti pattern.